[DCJ-15] Make DCJ team the default codeowners

okotsopoulos opened the pull request:
https://broadworkbench.atlassian.net/browse/DCJ-15

- Make @DataBiosphere/data-custodian-journeys (new Github team) the default codeowners
- Make @DataBiosphere/data-explorer-eng (existing Github team) codeowners of files in `snapshotbuilder` directories

### New Github Team

Here is the membership of our new team. Admins can modify membership and team settings:
<img width="1432" alt="Screenshot 2024-04-26 at 11 17 41 AM" src="https://github.com/DataBiosphere/jade-data-repo/assets/79769153/650edc6f-b4a7-4605-b5cc-82bf9090bb35">

I have made this team an [admin](https://github.com/DataBiosphere/jade-data-repo/settings/access?query=data-custodian-journeys) of this repository:
<img width="1398" alt="Screenshot 2024-04-26 at 11 33 46 AM" src="https://github.com/DataBiosphere/jade-data-repo/assets/79769153/8351f114-8c2b-4f41-8bb2-5ac6baf0ca46">

I also made the DE team a [writer](https://github.com/DataBiosphere/jade-data-repo/settings/access?query=data-explorer-eng) on this repository (required to make them codeowners):
<img width="1375" alt="Screenshot 2024-04-26 at 11 37 29 AM" src="https://github.com/DataBiosphere/jade-data-repo/assets/79769153/058d9c72-c910-4cd5-ba84-f804c3fbbe44">

### Code Review Settings

I have configured the new DCJ team so that 2 reviews are automatically requested when the team is a codeowner, rotating through reviewers via a load balancer strategy.  Sebastian and Gideon (our contractors) are members of this team but exempt from review requests:
<img width="1261" alt="Screenshot 2024-04-26 at 11 18 05 AM" src="https://github.com/DataBiosphere/jade-data-repo/assets/79769153/2e056666-8171-497a-9fe9-b0caff768741">

An example: I requested a review from the DCJ team on this open PR: https://github.com/DataBiosphere/jade-data-repo/pull/1647.  The load balancer automatically assigned Florian and Greg, and removed the request from the team.

More information on code review settings: https://docs.github.com/en/organizations/organizing-members-into-teams/managing-code-review-settings-for-your-team

### Scheduled Reminders

I have configured a [Slack scheduled reminder](https://github.com/orgs/DataBiosphere/teams/data-custodian-journeys/settings/reminders/72975) to #dsp-data-custodian-journeys on stale pull requests:
<img width="1344" alt="Screenshot 2024-04-26 at 11 35 21 AM" src="https://github.com/DataBiosphere/jade-data-repo/assets/79769153/0245988c-6e80-43ca-a01a-e3081a578ed7">

okotsopoulos committed:
[DCJ-15] Make @DataBiosphere/data-custodian-journeys team the default codeowners

And make @DataBiosphere/data-explorer-eng team codeowners of files in snapshotbuilder directories.

I have configured the new DCJ team so that 2 reviews are automatically requested when the team is a codeowner, rotating through reviewers via a load balancer strategy.  Sebastian and Gideon (our contractors) are members of this team but exempt from review requests.

okotsopoulos commented on .github/CODEOWNERS:
```diff
@@ -1,13 +1,13 @@
+# So if a pull request only touches files in a `/snapshotbuilder` directory
+# such as `/src/main/../snapshotbuilder` or `/src/test/../snapshotbuilder`,
+# only these owners will be requested to review.
+**/snapshotbuilder @DataBiosphere/data-explorer-eng
```
I'm open to expanding this somewhat, but I don't think we'll get it to be bulletproof: I know that cohort builder work can touch files outside of these directories.

fboulnois reviewed (approved):
Looks reasonable to me.

rushtong reviewed (approved):
This looks good to me, thank you 👍🏽

pshapiro4broad reviewed (changes requested):
Right now the work the data exploration team is often outside the snapshotbuilder directory, and there is no reason to think that this will change anytime in the next 6-8 months. Given that, it doesn't seem like a good idea to restrict code owners in this way.

Is there a reason why this change is necessary to do right now?

okotsopoulos commented:
> Right now the work the data exploration team is often outside the snapshotbuilder directory, and there is no reason to think that this will change anytime in the next 6-8 months. Given that, it doesn't seem like a good idea to restrict code owners in this way.
>
> Is there a reason why this change is necessary to do right now?

@pshapiro4broad , DCJ is stewarding this repository and we'd like to remove friction to reviewing our PRs.  The current codeowners structure doesn't allow for that and reflects an out-of-date team structure.  But, I'm open to doing something different:

- Make both DCJ and DE teams codeowners of the whole repo
- If a DE PR is raised, DCJ team members will have reviews requested but can remove themselves.  It's not a bad idea that we have visibility into what's going on in the repo we steward, though (we will be QA's first point of contact for release blockers stemming from this repo).
- If a DCJ PR is raised, we remove the DE review request.
- When snapshot builder work stabilizes, revisit codeowner structure.

I don't love this either but it may be the best compromise. Thoughts?  CC: @rushtong @fboulnois

fboulnois commented:
> @pshapiro4broad , DCJ is stewarding this repository and we'd like to remove friction to reviewing our PRs. The current codeowners structure doesn't allow for that and reflects an out-of-date team structure. But, I'm open to doing something different:
> 
>     * Make both DCJ and DE teams codeowners of the whole repo
> 
> I don't love this either but it may be the best compromise. Thoughts? CC: @rushtong @fboulnois

This seems reasonable to me.

rushtong commented:
> @pshapiro4broad , DCJ is stewarding this repository and we'd like to remove friction to reviewing our PRs. The current codeowners structure doesn't allow for that and reflects an out-of-date team structure. But, I'm open to doing something different:
> 
> * Make both DCJ and DE teams codeowners of the whole repo
> * If a DE PR is raised, DCJ team members will have reviews requested but can remove themselves.  It's not a bad idea that we have visibility into what's going on in the repo we steward, though (we will be QA's first point of contact for release blockers stemming from this repo).
> * If a DCJ PR is raised, we remove the DE review request.
> * When snapshot builder work stabilizes, revisit codeowner structure.
> 
> I don't love this either but it may be the best compromise. Thoughts? CC: @rushtong @fboulnois

I'm fine with this approach as well

pshapiro4broad commented:
> Make both DCJ and DE teams codeowners of the whole repo

That works, but is it better than adding the DCJ team members to https://github.com/orgs/DataBiosphere/teams/jadeteam?

My thinking is that `jadeteam` is already used in many other places, so it might be easier to continue to use it for now, as long as the repo has two teams that are actively working on it.

It also provides a way for other collaborators to be added and removed (as @dvoet is now) without adding them directly to the repo or to another team that they're not actually part of.

okotsopoulos commented:
> That works, but is it better than adding the DCJ team members to https://github.com/orgs/DataBiosphere/teams/jadeteam?

@pshapiro4broad  yes: the Jade team is a reflection of a team structure that does not exist anymore, and the split between teams makes it easier to request (or remove requests from) the appropriate team.

okotsopoulos commented:
@pshapiro4broad responding to your additions:

> My thinking is that jadeteam is already used in many other places, so it might be easier to continue to use it for now, as long as the repo has two teams that are actively working on it.

I personally find it easier to move away from this out-of-date team precisely because it's used in other repos, whose singular ownership will be more clear-cut (ex. I think TDR UI should be solely owned by DCJ team, even if DE team works in it from time to time).

> It also provides a way for other collaborators to be added and removed (as @dvoet is now) without adding them directly to the repo or to another team that they're not actually part of.

I don't know if Doug needed to be added to this team to work in the repo (maybe he was following our instructions rather than actually needing the membership, I can revisit those).

If / when that happens again we can investigate more closely and find a better solution: https://github.com/DataBiosphere/jade-data-repo/settings/access shows that `@DataBiosphere/broadwrite` should be able to write to the repo.

<img width="1344" alt="Screenshot 2024-04-26 at 1 29 24 PM" src="https://github.com/DataBiosphere/jade-data-repo/assets/79769153/e099bdb8-9ac9-4b7e-a2cd-3a6496cd6b85">

okotsopoulos committed:
PR feedback: DE team are also codeowners

Their snapshot builder work often touches code outside of /snapshotbuilder directories, and this is likely to be the case for the next 6 - 8 months.

okotsopoulos committed:
[docs] @DataBiosphere/broadwrite membership is usually sufficient

A DSP-er wanting to contribute to jade-data-repo is able to do so without needing to be a codeowner or a repository admin.

Our "Getting Started" guide has historically been tailored towards onboarding new members of a team stewarding the repo.

pshapiro4broad reviewed (approved):
Thanks!

okotsopoulos commented:
Note that I'll bypass branch protections to merge, because changes to the CODEOWNERS file do not trigger our test runs.
